### PR TITLE
Implemented series period detect method using pocketfft lib

### PR DESCRIPTION
**Pull request by bhavnajindal**
Implemented method seriesPeriodDetect to detect the period of  time series data. It makes use of the Fast Fourier Transform implementation available in [PocketFFT](https://gitlab.mpcdf.mpg.de/mtr/pocketfft/-/tree/cpp).

The result of this method could also be later used to decompose a series.

Input: An array of real numbers.
Returns : A real value.
Example use:

```sql
SELECT seriesPeriodDetect([139, 87, 110, 68, 54, 50, 51, 53, 133, 86, 141, 97, 156, 94, 149, 95, 140, 77, 61, 50, 54, 47, 133, 72, 152, 94, 148, 105, 162, 101, 160, 87, 63, 53, 55, 54, 151, 103, 189, 108, 183, 113, 175, 113, 178, 90, 71, 62, 62, 65, 165, 109, 181, 115, 182, 121, 178, 114, 170]);
-- returns: 14
```

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implemented series period detect method using FFT in pocketFFT lib

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit f9c384ffd388dea8bf41f261a535de790091adbd with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56171/f9c384ffd388dea8bf41f261a535de790091adbd/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

File: .gitmodules [hunk 1]
```diff
@@ -351,3 +351,7 @@
 [submodule "contrib/aklomp-base64"]
 	path = contrib/aklomp-base64
 	url = https://github.com/aklomp/base64.git
+[submodule "contrib/pocketfft"]
+	path = contrib/pocketfft
+	url = https://gitlab.mpcdf.mpg.de/mtr/pocketfft.git
+	branch = cpp
```
Comment (rschu1ze) [inline .gitmodules:357]: About `branch`, please see https://github.com/ClickHouse/ClickHouse/pull/56059/files

File: contrib/pocketfft-cmake/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+add_library(_pocketfft INTERFACE)
```
Comment (rschu1ze) [inline contrib/pocketfft-cmake/CMakeLists.txt:1]: Could we please make compilation dependent on `ENABLE_LIBRARIES`?

File: src/Functions/seriesPeriodDetect.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
+#pragma clang diagnostic push
```
Comment (rschu1ze) [inline src/Functions/seriesPeriodDetect.cpp:1]: Function documentation (subdirectory "docs/en/sql-reference/functions") is missing.

File: tests/queries/0_stateless/02813_series_period_detect.sql [hunk 4]
```diff
@@ -0,0 +1,3 @@
+SELECT seriesPeriodDetect([139, 87, 110, 68, 54, 50, 51, 53, 133, 86, 141, 97, 156, 94, 149, 95, 140, 77, 61, 50, 54, 47, 133, 72, 152, 94, 148, 105, 162, 101, 160, 87, 63, 53, 55, 54, 151, 103, 189, 108, 183, 113, 175, 113, 178, 90, 71, 62, 62, 65, 165, 109, 181, 115, 182, 121, 178, 114, 170]);
+SELECT seriesPeriodDetect([NULL, NULL, NULL]); -- { serverError 44 }
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02813_series_period_detect.sql:2]: Shortcut: `-- { serverError ILLEGAL_COLUMN }`

File: tests/queries/0_stateless/02813_series_period_detect.sql [hunk 5]
```diff
@@ -0,0 +1,3 @@
+SELECT seriesPeriodDetect([139, 87, 110, 68, 54, 50, 51, 53, 133, 86, 141, 97, 156, 94, 149, 95, 140, 77, 61, 50, 54, 47, 133, 72, 152, 94, 148, 105, 162, 101, 160, 87, 63, 53, 55, 54, 151, 103, 189, 108, 183, 113, 175, 113, 178, 90, 71, 62, 62, 65, 165, 109, 181, 115, 182, 121, 178, 114, 170]);
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02813_series_period_detect.sql:1]: Let's add tests also for floating point inputs (at the least - ideally, all other edge cases are also tested)

File: src/Functions/seriesPeriodDetect.cpp [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        || executeNumber<Int8>(src_data,  period)
+        || executeNumber<Int16>(src_data,  period)
+        || executeNumber<Int32>(src_data,  period)
+        || executeNumber<Int64>(src_data,  period)
+        || executeNumber<Float32>(src_data,  period)
+        || executeNumber<Float64>(src_data,  period)){
+            res_data[0] = period;
+            return res;
+        }
+        else
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp:61]:
```
if (!series_period) {
    // throw Exception(...);
}
return *series_period;
```

File: src/Functions/seriesPeriodDetect.cpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                    idx = i;
+                }
+            }
+
+            std::vector<double> xfreq(specLen);
+            double step = 0.5 / (specLen-1);
+            for(size_t i = 0; i < specLen; ++i)
+                xfreq[i] = i * step;
+
+            auto freq = xfreq[idx];
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp:110]: Can `freq` be `0`?

File: src/Functions/seriesPeriodDetect.cpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            for(size_t i = 1; i < specLen; ++i){
+                double magnitude =  sqrt(out[i].real() * out[i].real() + out[i].imag() * out[i].imag());
+                if(magnitude > maxMag){
+                    maxMag = magnitude;
+                    idx = i;
+                }
+            }
+
+            std::vector<double> xfreq(specLen);
+            double step = 0.5 / (specLen-1);
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp]: Can `specLen` be `1`? Please add tests for edge cases.

File: src/Functions/seriesPeriodDetect.cpp [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        }
+        else
+            throw Exception(
+                ErrorCodes::ILLEGAL_COLUMN,
+                "Illegal column {} of first argument of function {}",
+                    arguments[0].column->getName(), getName());
+    }
+
+    template <typename T>  
+    bool executeNumber(const IColumn & src_data, Float64 &period) const
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp]: Please document this function and the logic below for those not familiar with Fourier transform.
A reference to the algorithm would be great.

File: src/Functions/seriesPeriodDetect.cpp [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    {
+        return std::make_shared<DataTypeFloat64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t) const override
+    {
+        ColumnPtr array_ptr = arguments[0].column;
+        const ColumnArray * array = checkAndGetColumn<ColumnArray>(array_ptr.get());
+        const IColumn & src_data = array->getData();
+        //const ColumnArray::Offsets & offsets = array->getOffsets();
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp]: Not needed?

File: src/Functions/seriesPeriodDetect.cpp [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & /*arguments*/) const override
+    {
+        return std::make_shared<DataTypeFloat64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t) const override
+    {
+        ColumnPtr array_ptr = arguments[0].column;
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp:44]: Can `arguments` be empty?

File: src/Functions/seriesPeriodDetect.cpp [hunk 12]
```diff
@@ -135,7 +127,7 @@ class FunctionSeriesPeriodDetect : public IFunction
 
             auto freq = xfreq[idx];
 
-            period = (1/freq);
+            period = std::round(1/freq);
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp]: Please amend this change with the previous commit.

File: src/Functions/seriesPeriodDetect.cpp [hunk 13]
```diff
@@ -10,8 +10,7 @@
 #include <DataTypes/DataTypesNumber.h>
 #include <Columns/ColumnsNumber.h>
 #include <Columns/ColumnArray.h>
-
-
+#include <cmath>
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp:8]: Please amend this change with the previous commit.

**Review by jkartseva — CHANGES_REQUESTED** [review]
Can the commit with `contrib` and `src` changes be split, with the contrib changes in the first commit and the functional changes in a separate one? I would go even further and created a separate PR for the contrib and a submodule changes.

Also, why applying clang-format in a separate commit?

File: src/Functions/seriesPeriodDetect.cpp [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+            size_t len = src_vec.size();
+            std::vector<Float64> src(src_vec.begin(), src_vec.end());
+            std::vector<std::complex<double>> out((len/2)+1);
+
+            pocketfft::shape_t shape{static_cast<size_t>(len)};
+
+            pocketfft::shape_t axes;
+            for (size_t i=0; i<shape.size(); ++i)
+                axes.push_back(i);
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp]:
```
pocketfft::shape_t axes(shape.size());
std::iota(axes.begin(), axes.end(), 0);
```

https://en.cppreference.com/w/cpp/algorithm/iota

File: src/Functions/seriesPeriodDetect.cpp [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    template <typename T>  
+    bool executeNumber(const IColumn & src_data, Float64 &period) const
+    {
+        if(const ColumnVector<T> * src_data_concrete = checkAndGetColumn<ColumnVector<T>>(&src_data))
+        {
+            const PaddedPODArray<T> & src_vec = src_data_concrete->getData();
+
+            size_t len = src_vec.size();
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp:76]: Can `len` be `0`? Please test this edge case.

File: src/Functions/seriesPeriodDetect.cpp [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+            pocketfft::r2c(shape,
+                stride_src,
+                stride_out,
+                axes,
+                pocketfft::FORWARD,
+                src.data(),
+                out.data(),
+                static_cast<double>(1));
+
+            size_t specLen = (len/2 - (len%2 == 0 ? 1 : 0));  //removing the nyquist element when len is even
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp]: `size_t specLen  = (len - 1) / 2;`

**Review by jkartseva — COMMENTED** [review]
A couple of additional comments

File: src/Functions/seriesPeriodDetect.cpp (hunk 7 above)
Comment (bhavnajindal) [reply src/Functions/seriesPeriodDetect.cpp:110]: `freq` will never be zero as it depends on `idx` which starts from `1`

File: src/Functions/seriesPeriodDetect.cpp (hunk 11 above)
Comment (bhavnajindal) [reply src/Functions/seriesPeriodDetect.cpp:44]: Number of required arguments specified by `getNumberOfArguments()`. Added a test-case to verify.

File: src/Functions/seriesPeriodDetect.cpp [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+        /* const ColumnPtr & col = arguments[0].column;
+        const ColumnVector<Int64> * src_col = checkAndGetColumn<ColumnVector<Int64>>(col.get());
+        const PaddedPODArray<Int64> & values = src_col->getData(); */
+
+        
+
+        auto res = ColumnFloat64::create(1);
+        auto & res_data = res->getData();
+
+        Float64 period;
```
[the remark this answers is not in the document]
Comment (bhavnajindal) [reply src/Functions/seriesPeriodDetect.cpp:51]: This could be good refactoring and I earlier did consider doing it this way. But then I decided to keep it similar to other implementations of `IFunction`  such as [src/Functions/array/arrayDistinct.cpp](url) and [src/Functions/array/arrayReverse.cpp](url)

**Comment by bhavnajindal** [comment]
Updated the PR with some edge test cases, documentation and other requested changes.

File: src/Functions/seriesPeriodDetect.cpp [hunk 18]
```diff
@@ -69,51 +78,53 @@ class FunctionSeriesPeriodDetect : public IFunction
...
-        {
-            const PaddedPODArray<T> & src_vec = src_data_concrete->getData();
+        const ColumnVector<T> * src_data_concrete = checkAndGetColumn<ColumnVector<T>>(&src_data);
+        if (!src_data_concrete)
+            return false;
+
+        const PaddedPODArray<T> & src_vec = src_data_concrete->getData();
+
+        size_t len = src_vec.size();
+        if (len < 4)
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetect.cpp:88]: Can there be a case with too many arguments?

**Review by jkartseva — CHANGES_REQUESTED** [review]
Since my previous request to organize this PR into atomic commits was not addressed, I'm asking again.
* Please separate `contib` changes that add the new dependency from the functional changes.
* Please merge bug- and style fixes into the commit they address. This will simplify the review and the following maintenance process by a great deal. Thank you

File: .gitmodules [hunk 19]
```diff
@@ -351,3 +351,6 @@
 [submodule "contrib/aklomp-base64"]
 	path = contrib/aklomp-base64
 	url = https://github.com/aklomp/base64.git
+[submodule "contrib/pocketfft"]
+	path = contrib/pocketfft
+	url = https://gitlab.mpcdf.mpg.de/mtr/pocketfft.git
```
Comment (rschu1ze) [inline .gitmodules:356]: Looks like this repo is pretty much dead, and that development moved on to https://github.com/mreineck/pocketfft. Let's reference the other repo.

File: docs/en/sql-reference/functions/time-series-functions.md [hunk 20]
```diff
@@ -0,0 +1,45 @@
+---
+slug: /en/sql-reference/functions/time-series-functions
+sidebar_position: 140
```
Comment (rschu1ze) [inline docs/en/sql-reference/functions/time-series-functions.md:3]: 140 determines the position in the sidebar (e.g. [here](https://clickhouse.com/docs/en/sql-reference/functions/time-window-functions)). To retain sorting, this page should go between "Strings" (170) and "Time Window" (175). Let's use 172.

File: docs/en/sql-reference/functions/time-series-functions.md [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+sidebar_label: Time Series
+---
+
+# Time Series Functions
+
+Below functions are used for time series anlaysis.
```
Comment (rschu1ze) [inline docs/en/sql-reference/functions/time-series-functions.md:9]: analysis

File: docs/en/sql-reference/functions/time-series-functions.md [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+## seriesPeriodDetect
```
Comment (rschu1ze) [inline docs/en/sql-reference/functions/time-series-functions.md:11]: You are more of an expert with this but are there other viable algorithms to detect periods in time series? E.g. algorithms with different trade-offs? E.g. (*) ... In that case, someone else may want to implement them later, and it would be more consistent to have the algorithms name appended to the function name, e.g. `seriesPeriodDetectFFT()`. (But I am really not sure here, kindly check)

(*)
https://www.cs.purdue.edu/homes/ake/pub/Tkde-0129-0504-1.pdf
https://stats.stackexchange.com/questions/16117/what-method-can-be-used-to-detect-seasonality-in-data
https://project.inria.fr/aaltd19/files/2019/08/AALTD_19_Boussard.pdf
https://www.sciencedirect.com/science/article/abs/pii/S0031320323000560

File: src/CMakeLists.txt [hunk 23]
```diff
@@ -410,6 +410,8 @@ dbms_target_link_libraries(PRIVATE ch_contrib::zstd)
 target_link_libraries (clickhouse_common_io PUBLIC ch_contrib::zstd)
 target_link_libraries (clickhouse_common_io PUBLIC ch_contrib::xz)
 
+target_link_libraries(clickhouse_common_io PUBLIC ch_contrib::pocketfft)
```
Comment (rschu1ze) [inline src/CMakeLists.txt:413]: Needs conditional wrapping:

```cmake
if (TARGET ch_contrib::pocketfft)
    target_link_libraries(clickhouse_common_io PUBLIC ch_contrib::pocketfft)
endif ()
```

(otherwise builds with `-DENABLE_LIBRARIES=0` will break)

File: src/Functions/CMakeLists.txt [hunk 24]
```diff
@@ -41,6 +41,7 @@ list (APPEND PRIVATE_LIBS
         boost::filesystem
         divide_impl
         ch_contrib::xxHash
+        ch_contrib::pocketfft
```
Comment (rschu1ze) [inline src/Functions/CMakeLists.txt:44]: I am pretty sure that we don't need that.

File: src/Functions/seriesPeriodDetect.cpp [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int BAD_ARGUMENTS;
+extern const int ILLEGAL_COLUMN;
+}
+
+/*Detect Period in time series data using FFT.
+ * FFT - Fast Fourier transform (https://en.wikipedia.org/wiki/Fast_Fourier_transform)
```
Comment (rschu1ze) [inline src/Functions/seriesPeriodDetect.cpp:24]: Since this is about the algorithms itself and not details of the C++ implementation, I suppose we should have that also in the user-facing documentation (markdown file).

File: src/Functions/seriesPeriodDetect.cpp [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+public:
+    static constexpr auto name = "seriesPeriodDetect";
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionSeriesPeriodDetect>(); }
+
+    std::string getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool isVariadic() const override { return false; }
```
Comment (rschu1ze) [inline src/Functions/seriesPeriodDetect.cpp:42]: (that's the default in the base class `IFunction` --> method can be omitted)

**Comment by bhavnajindal** [comment]
> Since my previous request to organize this PR into atomic commits was not addressed, I'm asking again.
> 
> * Please separate `contib` changes that add the new dependency from the functional changes.
> * Please merge bug- and style fixes into the commit they address. This will simplify the review and the following maintenance process by a great deal. Thank you

Thank you for mentioning this again. I missed out on your previous comment regarding this.

Since I have made a lot of commits in this PR, it would be quite complicated to use `git commit --amend` to separate the changes and squash functional changes. Would it be good idea to raise a new PR with atomic commits after we are mostly done with the review of functional changes? Kindly recommend.

File: docs/en/sql-reference/functions/time-series-functions.md (hunk 22 above)
Comment (bhavnajindal) [reply docs/en/sql-reference/functions/time-series-functions.md:11]: You are right, there are other are other algorithms. I will modify the method name.

File: src/Functions/seriesPeriodDetect.cpp (hunk 18 above)
Comment (bhavnajindal) [reply src/Functions/seriesPeriodDetect.cpp:88]: It should be able to accept the maximum array that can be passed to CH functions. Overall, the method accepts only one 1 argument (specified in `getNumberOfArguments`).

**Comment by jkartseva** [comment]
@bhavnajindal 

It's okay to proceed with the current commit structure, given the number of review comments. No need to open a separate PR, please address the existing comments concerning the functional part, and it should be okay to merge as is.

**Comment by bhavnajindal** [comment]
Thank you @jkartseva 
I have addressed the existing review comments.

**Review by rschu1ze — COMMENTED** [review]
Thanks, no further objections from my side, @jkartseva may do the final approve.

File: tests/queries/0_stateless/02813_series_period_detect.sql [hunk 27]
```diff
@@ -0,0 +1,9 @@
+SELECT seriesPeriodDetectFFT([139, 87, 110, 68, 54, 50, 51, 53, 133, 86, 141, 97, 156, 94, 149, 95, 140, 77, 61, 50, 54, 47, 133, 72, 152, 94, 148, 105, 162, 101, 160, 87, 63, 53, 55, 54, 151, 103, 189, 108, 183, 113, 175, 113, 178, 90, 71, 62, 62, 65, 165, 109, 181, 115, 182, 121, 178, 114, 170]);
```
Comment (jkartseva) [inline tests/queries/0_stateless/02813_series_period_detect.sql:1]: Add a test case where all values are equal?

File: src/Functions/seriesPeriodDetectFFT.cpp [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+
+        pocketfft::r2c(shape, stride_src, stride_out, axes, pocketfft::FORWARD, src.data(), out.data(), static_cast<double>(1));
+
+        size_t specLen = (len - 1) / 2; //removing the nyquist element when len is even
+
+        double maxMag = 0;
+        size_t idx = 1;
+        for (size_t i = 1; i < specLen; ++i)
+        {
+            double magnitude = sqrt(out[i].real() * out[i].real() + out[i].imag() * out[i].imag());
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetectFFT.cpp:116]: `sqrt` can be omitted because `maxMag` is used only to get the frequency index. The square of the magnitude (`maxMagnituteSquare`?) works just fine for the purpose.

File: src/Functions/seriesPeriodDetectFFT.cpp [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+        axes.reserve(shape.size());
+        for (size_t i = 0; i < shape.size(); ++i)
+            axes.push_back(i);
+
+        pocketfft::stride_t stride_src{sizeof(double)};
+        pocketfft::stride_t stride_out{sizeof(std::complex<double>)};
+
+        pocketfft::r2c(shape, stride_src, stride_out, axes, pocketfft::FORWARD, src.data(), out.data(), static_cast<double>(1));
+
+        size_t specLen = (len - 1) / 2; //removing the nyquist element when len is even
```
Comment (jkartseva) [inline src/Functions/seriesPeriodDetectFFT.cpp:110]: We should be consisted with using either snake case (e.g. `stride_src`) or camel case (e.g. `specLen`) for variable names.

**Review by jkartseva — COMMENTED** [review]
Thank you for addressing the comments @bhavnajindal . I have a few minor comments and will approve if CIs don't fail because of the `pocketfft` dependency.

It the run associated with this PR the headers are not found:

```
2023-11-07 21:17:06 /ClickHouse/src/Functions/seriesPeriodDetectFFT.cpp:8:10: fatal error: 'pocketfft_hdronly.h' file not found
2023-11-07 21:17:06     8 | #include "pocketfft_hdronly.h"
2023-11-07 21:17:06       |          ^~~~~~~~~~~~~~~~~~~~~
2023-11-07 21:17:06 1 error generated.
```

https://github.com/ClickHouse/ClickHouse/actions/runs/6789642655/job/18457731024?pr=56171#step:6:4504

Could you please take a look?

**Comment by jkartseva** [comment]
I believe CIs have failed because of the spelling check:
```
Nov 09 22:11:08 Check docs spelling
====== /ClickHouse/docs/en/sql-reference/functions/time-series-functions.md ======
seriesPeriodDetectFFT
====== Errors found ======
To exclude some words add them to the dictionary file "/ClickHouse/utils/check-style/aspell-ignore/en/aspell-dict.txt"
You can also run ./check-doc-aspell -i to see the errors interactively and fix them or add to the dictionary file
```

https://github.com/ClickHouse/ClickHouse/actions/runs/6817836954/job/18542431597?pr=56171

File: src/Functions/seriesPeriodDetectFFT.cpp [hunk 30]
```diff
@@ -0,0 +1,167 @@
...
+
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        const DataTypeArray * array_type = checkAndGetDataType<DataTypeArray>(arguments[0].get());
```
Comment (rschu1ze) [inline src/Functions/seriesPeriodDetectFFT.cpp:61]: The "modern" way to do it would be `FunctionArgumentDescriptors` and `validateFunctionArgumentTypes()`, see its usage in Functions/makeDate.cpp.

**Comment by bhavnajindal** [comment]
Hi @jkartseva - Looks like the CI is failing due to undocumented 'seriesPeriodDetectFFT()' (https://github.com/ClickHouse/ClickHouse/actions/runs/6829158775/job/18582481507?pr=56171#step:6:2368). Where could I add the documentation to fix this test.

**Comment by jkartseva** [comment]
@bhavnajindal 

I have a broader question — sorry I'm asking this after several review iterations, not beforehand. Could you please elaborate on the reasoning behind picking pocketfft as the library for the FFT? There are several projects on GitHub, why have you chosen pocketFFT in the first place? Ideally, we are looking for a library that is actively maintained.
Side note: this is [not the case](https://github.com/anthonix/ffts/issues/92) for the [top suggestion](https://github.com/search?q=fft+language%3AC%2B%2B&type=repositories&l=C%2B%2B) for Fast Fourier Transform.

**Comment by bhavnajindal** [comment]
@jkartseva 
It's true that there are several libraries available for FFT. The most recommend being FFTW(https://www.fftw.org) and also my first choice. But, FFTW has a GPL license leading me to look for other libraries.
I chose to go with PocketFFT because, 

- it's  BSD license.
- Has really good performance benchmarks against other libs (https://github.com/project-gemmi/benchmarking-fft/)
- Has a pure C++ implementation compared to some other fast libraries which are only available in C.
- It supports vectorization.

**Comment by rschu1ze** [comment]
Yes, GPL is a no-no, see below ... only LGPL would be acceptable.

Numpy and Eigen are very established libraries, that makes a strong argument for PocketFFT.

Green light from my side.

```sql
SELECT
    library_name,
    license_type
FROM system.licenses

Query id: 26e808b1-398a-4236-afa9-cc9d42aa499e

┌─library_name────────┬─license_type─────────┐
│ poco                │ Boost                │
│ AMQP-CPP            │ Apache               │
│ FP16                │ MIT                  │
│ NuRaft              │ Apache               │
│ SimSIMD             │ Apache               │
│ abseil-cpp          │ Apache               │
│ aklomp-base64       │ BSD 2-clause         │
│ annoy               │ Apache               │
│ arrow               │ Apache               │
│ avro                │ Apache               │
│ aws-c-auth          │ Apache               │
│ aws-c-cal           │ Apache               │
│ aws-c-common        │ Apache               │
│ aws-c-compression   │ Apache               │
│ aws-c-event-stream  │ Apache               │
│ aws-c-http          │ Apache               │
│ aws-c-io            │ Apache               │
│ aws-c-mqtt          │ Apache               │
│ aws-c-s3            │ Apache               │
│ aws-c-sdkutils      │ Apache               │
│ aws-checksums       │ Apache               │
│ aws-crt-cpp         │ Apache               │
│ aws-s2n-tls         │ Apache               │
│ aws                 │ Apache               │
│ azure               │ MIT                  │
│ boost               │ Boost                │
│ boringssl           │ BSD                  │
│ brotli              │ MIT                  │
│ bzip2               │ bzip2                │
│ c-ares              │ HPND                 │
│ capnproto           │ MIT                  │
│ cassandra           │ Apache               │
│ cctz                │ Apache               │
│ cityhash102         │ MIT                  │
│ cld2                │ Apache               │
│ corrosion           │ MIT                  │
│ cppkafka            │ BSD 2-clause         │
│ crc32-s390x         │ Apache               │
│ crc32-vpmsum        │ Apache               │
│ croaring            │ Apache               │
│ curl                │ MIT/curl             │
│ cyrus-sasl          │ BSD 2-clause         │
│ datasketches-cpp    │ Apache               │
│ double-conversion   │ BSD 3-clause         │
│ dragonbox           │ Boost                │
│ fast_float          │ Apache               │
│ fastops             │ MIT                  │
│ flatbuffers         │ Apache               │
│ fmtlib              │ MIT                  │
│ google-benchmark    │ Apache               │
│ google-protobuf     │ BSD 3-clause         │
│ googletest          │ BSD 3-clause         │
│ grpc                │ Apache               │
│ h3                  │ Apache               │
│ hive-metastore      │ Apache               │
│ icu                 │ Public Domain        │
│ icudata             │ Public Domain        │
│ idxd-config         │ Unknown              │
│ isa-l               │ BSD 3-clause         │
│ jemalloc            │ BSD 2-clause         │
│ krb5                │ Unknown              │
│ lemmagen-c          │ LGPL                 │
│ libarchive          │ Apache               │
│ libbcrypt           │ Unknown              │
│ libcpuid            │ BSD 2-clause         │
│ libdivide           │ Boost                │
│ libfarmhash         │ MIT                  │
│ libfiu              │ Public Domain        │
│ libgsasl            │ LGPL                 │
│ libhdfs3            │ Apache               │
│ libmetrohash        │ Apache               │
│ libpq               │ PostgreSQL           │
│ libpqxx             │ BSD 3-clause         │
│ libprotobuf-mutator │ Apache               │
│ librdkafka          │ Apache               │
│ libssh              │ LGPL                 │
│ libstemmer_c        │ BSD 3-clause         │
│ libunwind           │ Apache               │
│ liburing            │ MIT                  │
│ libuv               │ Public Domain        │
│ libxml2             │ MIT                  │
│ llvm-project        │ Apache               │
│ lz4                 │ BSD                  │
│ magic_enum          │ MIT                  │
│ mariadb-connector-c │ LGPL                 │
│ miniselect          │ Boost                │
│ minizip-ng          │ zLib                 │
│ morton-nd           │ MIT                  │
│ msgpack-c           │ Boost                │
│ murmurhash          │ Public Domain        │
│ nanodbc             │ MIT                  │
│ nats-io             │ Apache               │
│ openldap            │ OpenLDAP Version 2.8 │
│ openssl             │ Apache               │
│ orc                 │ Apache               │
│ pdqsort             │ zLib                 │
│ qpl                 │ MIT                  │
│ rapidjson           │ BSD                  │
│ re2                 │ BSD 3-clause         │
│ replxx              │ BSD 3-clause         │
│ robin-map           │ MIT                  │
│ rocksdb             │ BSD 3-clause         │
│ s2geometry          │ Apache               │
│ sentry-native       │ MIT                  │
│ simdjson            │ Apache               │
│ snappy              │ Public Domain        │
│ sparsehash-c11      │ BSD 3-clause         │
│ sqlite-amalgamation │ BSD 3-clause         │
│ sysroot             │ LGPL                 │
│ thrift              │ Apache               │
│ unixodbc            │ LGPL                 │
│ usearch             │ Apache               │
│ vectorscan          │ Boost                │
│ wyhash              │ Public Domain        │
│ xxHash              │ BSD                  │
│ xz                  │ LGPL                 │
│ yaml-cpp            │ MIT                  │
│ zlib-ng             │ zLib                 │
│ zstd                │ BSD                  │
└─────────────────────┴──────────────────────┘
```

**Review by jkartseva — APPROVED** [review]

**Comment by jkartseva** [comment]
I believe 100 samples is too few. Here's how the function performs on `500`: https://pastila.nl/?0060ca67/e20d1379f01257644ad06d995fe0ad29#QQS1Q2KFVvAprY71bjKalw==
The result is close to the anticipated period of `20π`.

I've experimented with `numpy` I'm more familiar with and it demonstrates the same behavior on a few samples:

100 samples: https://pastila.nl/?040543da/2c1e955fb7489b467b3a1d9b423e152f#SZspwUxGOYXpUWL7O5rc2A==
1000 samples: https://pastila.nl/?03b155d6/8f40a91925f43270909bf1a2f740c02a#qq4OE8c7G8zwMPWeYf1T/Q==

**Comment by bhavnajindal** [comment]
As noted by @jkartseva, 100 samples are too few when the expected period is around 62. For better results, it's best to have number of samples in the multiples of the period. This would  allow FFT to better capture the frequency distribution in data and hence be able to accurately generate period.

Some of the runs shared by @alexey-milovidov fails due to the exception during `dry run`. I have now overridden the `executeImplDryRun` method in my local environment and tested. It works as expected: https://pastila.nl/?00408b98/001af2b5bc6810d83744cfcbbb4f2602#CqC+prTXrqQWcONWxWpPJA==   

I would raise a PR with this fix.
